File: fluidresizelayout/src/main/java/me/saket/fluidresize/KeyboardVisibilityDetector.java
package me.saket.fluidresize;

import android.view.ViewTreeObserver;

final class KeyboardVisibilityDetector {

    static void listen(ActivityViewHolder viewHolder, KeyboardVisibilityChangedListener listener) {
        Detector detector = new Detector(viewHolder, listener);
        viewHolder.getNonResizableLayout().getViewTreeObserver().addOnPreDrawListener(detector);
        viewHolder.onDetach(() -> viewHolder.getNonResizableLayout().getViewTreeObserver().removeOnPreDrawListener(detector));
    }

    private static final class Detector implements ViewTreeObserver.OnPreDrawListener {

        private final ActivityViewHolder viewHolder;
        private final KeyboardVisibilityChangedListener listener;

        private int previousHeight = -1;

        Detector(ActivityViewHolder viewHolder, KeyboardVisibilityChangedListener listener) {
            this.viewHolder = viewHolder;
            this.listener = listener;
        }

        @Override
        public boolean onPreDraw() {
            return !detect();
        }

        private boolean detect() {
            int contentHeight = viewHolder.getResizableLayout().getHeight();
            if (contentHeight == previousHeight) {
                return false;
            }

            if (previousHeight != -1) {
                int statusBarHeight = viewHolder.getResizableLayout().getTop();
                boolean isKeyboardVisible = contentHeight < viewHolder.getNonResizableLayout().getHeight() - statusBarHeight;

                listener.onKeyboardVisibilityChanged(new KeyboardVisibilityChangedListener.KeyboardVisibilityChangedEvent(
                        isKeyboardVisible, contentHeight, previousHeight
                ));
            }

            previousHeight = contentHeight;
            return true;
        }
    }
}

File: fluidresizelayout/src/main/java/me/saket/fluidresize/KeyboardVisibilityDetector.kt
package me.saket.fluidresize

import android.view.ViewTreeObserver

object KeyboardVisibilityDetector {

  fun listen(viewHolder: ActivityViewHolder, listener: (KeyboardVisibilityChanged) -> Unit) {
    val detector = Detector(viewHolder, listener)
    viewHolder.nonResizableLayout.viewTreeObserver.addOnPreDrawListener(detector)
    viewHolder.onDetach {
      viewHolder.nonResizableLayout.viewTreeObserver.removeOnPreDrawListener(detector)
    }
  }

  private class Detector(
      val viewHolder: ActivityViewHolder,
      val listener: (KeyboardVisibilityChanged) -> Unit
  ) : ViewTreeObserver.OnPreDrawListener {

    private var previousHeight: Int = -1

    override fun onPreDraw(): Boolean {
      val detected = detect()

      // The layout flickers for a moment, usually on the first
      // animation. Intercepting this pre-draw seems to solve the problem.
      return detected.not()
    }

    private fun detect(): Boolean {
      val contentHeight = viewHolder.resizableLayout.height
      if (contentHeight == previousHeight) {
        return false
      }

      if (previousHeight != -1) {
        val statusBarHeight = viewHolder.resizableLayout.top
        val isKeyboardVisible = contentHeight < viewHolder.nonResizableLayout.height - statusBarHeight

        listener(KeyboardVisibilityChanged(
            visible = isKeyboardVisible,
            contentHeight = contentHeight,
            contentHeightBeforeResize = previousHeight))
      }

      previousHeight = contentHeight
      return true
    }
  }
}

